gix-config/src/tests.rs

```diff
@@ -0,0 +1,67 @@
+mod impls {
+    use bstr::ByteSlice;
+
+    use crate::Key;
+
+    #[test]
+    fn key_impl_for_str() {
+        assert_eq!("", "".section_name());
+        assert_eq!(None, "".subsection_name());
+        assert_eq!("", "".name());
+
+        assert_eq!("", "foo".section_name());
+        assert_eq!(None, "foo".subsection_name());
+        assert_eq!("foo", "foo".name());
```

Comment:
Related to my earlier comment, this is a good example on where the API of `Key` isn't currently suitable to capture what a `Key` is. Top-level keys like `"foo"` are not valid, they need a section name, and empty section names are not permitted (nor are empty keys).

Thus I think it's important to be able to express a parsing failure or 'invalid key' as part of the `Key` trait (or `ParseKey` trait as proposed).